stream.go

```diff
@@ -102,26 +105,36 @@ func NewClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, meth
 	if err != nil {
 		return nil, toRPCErr(err)
 	}
+	if EnableTracing {
+		trInfo.tr = trace.New("Sent."+methodFamily(desc.StreamName), desc.StreamName)
+		trInfo.firstLine.client = true
+		if deadline, ok := ctx.Deadline(); ok {
+			trInfo.firstLine.deadline = deadline.Sub(time.Now())
+		}
+		trInfo.tr.LazyLog(&trInfo.firstLine, false)
+	}
 	s, err := t.NewStream(ctx, callHdr)
 	if err != nil {
 		return nil, toRPCErr(err)
 	}
 	return &clientStream{
-		t:     t,
-		s:     s,
-		p:     &parser{s: s},
-		desc:  desc,
-		codec: cc.dopts.codec,
+		t:         t,
+		s:         s,
+		p:         &parser{s: s},
+		desc:      desc,
+		codec:     cc.dopts.codec,
+		traceInfo: trInfo,
```

Comment:
On Wed, Jun 17, 2015 at 7:58 PM, David Symonds <notifications@github.com>
wrote:

> In stream.go
> <https://github.com/grpc/grpc-go/pull/229#discussion_r32695701>:
>
> >  	s, err := t.NewStream(ctx, callHdr)
> >  	if err != nil {
> >  		return nil, toRPCErr(err)
> >  	}
> >  	return &clientStream{
> > -		t:     t,
> > -		s:     s,
> > -		p:     &parser{s: s},
> > -		desc:  desc,
> > -		codec: cc.dopts.codec,
> > +		t:         t,
> > +		s:         s,
> > +		p:         &parser{s: s},
> > +		desc:      desc,
> > +		codec:     cc.dopts.codec,
> > +		traceInfo: trInfo,
>
> this is incorrect. This is making a copy of the traceInfo struct, but the
> trace has a pointer to the original, so any changes to firstLine after this
> will not be reflected, for instance. It'll also pin both bits of memory.
>
I agree this needs to be changed. But the only reason is that it introduces
unnecessary copying. This is the exit point of the entire function so that
I do not understand what you meant by "any changes to firstLine after this
will not be reflected". Do I miss anything here?

> Either clientStream needs to be allocated higher up (probably the simplest
> option), or it needs to have a *traceInfo field instead.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/229/files#r32695701>.
>